libraries/chain/resource_limits.cpp

```diff
@@ -441,4 +441,136 @@ account_resource_limit resource_limits_manager::get_account_net_limit_ex( const
    return arl;
 }
 
+void resource_limits_manager::update_proxied(int64_t now, symbol purpose_symbol, const account_name& account, int64_t frame_length, bool force) {
+    const auto& agents_idx = _db.get_mutable_index<stake_agent_index>().indices().get<stake_agent_object::by_key>();
+    const auto& grants_idx = _db.get_mutable_index<stake_grant_index>().indices().get<stake_grant_object::by_key>();
+    update_proxied_traversal(now, purpose_symbol, agents_idx, grants_idx, get_agent(purpose_symbol, agents_idx, account), 
+        frame_length, force);
+}
+
+void resource_limits_manager::recall_proxied(int64_t now, account_name grantor_name, account_name agent_name, 
+                    symbol_code token_code, symbol_code purpose_code, int16_t pct) {
+                        
+    EOS_ASSERT(1 <= pct && pct <= config::_100percent, transaction_exception, "pct must be between 0.01% and 100% (1-10000)");
+    const auto* param = _db.find<stake_param_object, by_id>(token_code.value);
+    EOS_ASSERT(param, transaction_exception, "no staking for token");
+    auto purpose_symbol = param->get_purpose_symbol(token_code, purpose_code);
+    
+    const auto& agents_idx = _db.get_mutable_index<stake_agent_index>().indices().get<stake_agent_object::by_key>();
+    const auto& grants_idx = _db.get_mutable_index<stake_grant_index>().indices().get<stake_grant_object::by_key>();
+
+    auto grantor_as_agent = get_agent(purpose_symbol, agents_idx, grantor_name);
+    
+    update_proxied_traversal(now, purpose_symbol, agents_idx, grants_idx, grantor_as_agent, param->frame_length, false);
+    
+    int64_t amount = 0;
+    auto grant_itr = grants_idx.lower_bound(grant_key(purpose_symbol, grantor_name));
+    while ((grant_itr != grants_idx.end()) &&
+           (grant_itr->purpose_id   == purpose_symbol.decimals()) &&
+           (grant_itr->token_code   == purpose_symbol.to_symbol_code()) &&
+           (grant_itr->grantor_name == grantor_name))
+    {
+        if (grant_itr->agent_name == agent_name) {
+            auto to_recall = static_cast<int64_t>((static_cast<int128_t>(grant_itr->share) * pct) / config::_100percent);
+            amount = recall_proxied_traversal(purpose_symbol, agents_idx, grants_idx, grant_itr->agent_name, to_recall, grant_itr->break_fee);
```

Comment:
looks like `amount` may be overwritten several times. is it ok?
or if there is "at-most-once" case, then `break` will help to exit `while` early

Reply:
{goal identifier, token code, provider name, agent name} is unique here, so amount will be calculated only once.
I'll add break, thx.